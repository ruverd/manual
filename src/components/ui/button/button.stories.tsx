import type { Meta, StoryObj } from '@storybook/react';

import { Button } from './button';

const meta = {
  title: 'UI/Button',
  component: Button,
  parameters: {
    layout: 'centered'
  },
  tags: ['autodocs']
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    children: 'Button Text'
  }
};

export const AsLink: Story = {
  args: {
    asChild: true,
    children: <a href="#">Link Button</a>
  }
};

export const Disabled: Story = {
  args: {
    children: 'Disabled Button',
    disabled: true
  }
};

export const WithIcon: Story = {
  args: {
    children: (
      <>
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="24"
          height="24"
          viewBox="0 0 24 24"
          fill="none"
          stroke="currentColor"
          strokeWidth="2"
          strokeLinecap="round"
          strokeLinejoin="round"
        >
          <path d="M5 12h14" />
          <path d="m12 5 7 7-7 7" />
        </svg>
        Button with Icon
      </>
    )
  }
};
