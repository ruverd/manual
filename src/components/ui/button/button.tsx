import { Slot } from '@radix-ui/react-slot';
import * as React from 'react';

import { cn } from '@/lib/utils';

import type { ButtonProps } from './button.types';

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, asChild = false, ...props }, ref) => {
    const Comp = asChild ? Slot : 'button';

    return (
      <Comp
        className={cn(
          'inline-flex h-8 items-center justify-center whitespace-nowrap bg-primary px-[30px] py-[15px] text-[10px] font-bold uppercase tracking-[0.15em] text-primary-foreground transition-colors hover:bg-primary/90 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-20 md:h-[45px] [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0',
          className
        )}
        ref={ref}
        {...props}
      />
    );
  }
);

Button.displayName = 'Button';

export { Button };
