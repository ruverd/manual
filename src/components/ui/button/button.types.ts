export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  asChild?: boolean;
}
