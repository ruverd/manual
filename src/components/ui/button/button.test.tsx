import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { describe, expect, it, vi } from 'vitest';

import { Button } from './button';

const defaultProps = {
  children: 'Click me',
  className: 'custom-class',
  onClick: vi.fn()
};

const renderComponent = (props = {}) => {
  return render(<Button {...defaultProps} {...props} />);
};

describe('Button', () => {
  it('should render button with correct text', () => {
    renderComponent();

    expect(screen.getByRole('button')).toHaveTextContent('Click me');
  });

  it('should apply custom className along with default styles', () => {
    renderComponent();

    const button = screen.getByRole('button');
    expect(button).toHaveClass('custom-class');
    expect(button).toHaveClass('inline-flex');
    expect(button).toHaveClass('bg-primary');
  });

  it('should handle click events', async () => {
    const user = userEvent.setup();
    const onClick = vi.fn();
    render(<Button onClick={onClick}>Click me</Button>);

    const button = screen.getByRole('button');
    await user.click(button);

    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('should render as a different element when asChild is true', () => {
    render(
      <Button asChild>
        <a href="/test">Link Button</a>
      </Button>
    );

    const link = screen.getByRole('link');
    expect(link).toBeInTheDocument();
    expect(link).toHaveTextContent('Link Button');
    expect(link).toHaveAttribute('href', '/test');
  });

  it('should be disabled when disabled prop is true', () => {
    render(<Button disabled>Disabled Button</Button>);

    const button = screen.getByRole('button');
    expect(button).toBeDisabled();
    expect(button).toHaveClass('disabled:opacity-20');
  });

  it('should forward ref correctly', () => {
    const ref = React.createRef<HTMLButtonElement>();
    render(<Button ref={ref}>Button with ref</Button>);

    expect(ref.current).toBeInstanceOf(HTMLButtonElement);
  });

  it('should match snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when disabled', () => {
    const { container } = render(<Button disabled>Disabled Button</Button>);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with asChild', () => {
    const { container } = render(
      <Button asChild>
        <a href="/test">Link Button</a>
      </Button>
    );
    expect(container).toMatchSnapshot();
  });
});
