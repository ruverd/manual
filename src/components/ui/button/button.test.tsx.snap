// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should match snapshot 1`] = `
<div>
  <button
    class="inline-flex h-8 items-center justify-center whitespace-nowrap bg-primary px-[30px] py-[15px] text-[10px] font-bold uppercase tracking-[0.15em] text-primary-foreground transition-colors hover:bg-primary/90 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-20 md:h-[45px] [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 custom-class"
  >
    Click me
  </button>
</div>
`;

exports[`Button > should match snapshot when disabled 1`] = `
<div>
  <button
    class="inline-flex h-8 items-center justify-center whitespace-nowrap bg-primary px-[30px] py-[15px] text-[10px] font-bold uppercase tracking-[0.15em] text-primary-foreground transition-colors hover:bg-primary/90 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-20 md:h-[45px] [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0"
    disabled=""
  >
    Disabled Button
  </button>
</div>
`;

exports[`Button > should match snapshot with asChild 1`] = `
<div>
  <a
    class="inline-flex h-8 items-center justify-center whitespace-nowrap bg-primary px-[30px] py-[15px] text-[10px] font-bold uppercase tracking-[0.15em] text-primary-foreground transition-colors hover:bg-primary/90 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-20 md:h-[45px] [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0"
    href="/test"
  >
    Link Button
  </a>
</div>
`;
